src/tests/tracker.rs

```diff
@@ -0,0 +1,75 @@
+use std::sync::{atomic::AtomicBool, Arc};
+
+use crate::{Datadog, TrackerConfiguration};
+
+use super::mocks::{expect_event, expect_incr, MockClient};
+
+#[test]
+pub fn no_actions_tracker_does_nothing() {
+    let mut mock = MockClient::new();
+    let tracker_config = TrackerConfiguration::new().with_threshold(5);
+    // Expect 100 increment calls
+    for i in 0..100 {
+        mock = expect_incr(mock, "test", vec![format!("{}", i)]);
+    }
+    let dd = Datadog::new(mock, true, tracker_config);
+    for i in 0..100 {
+        dd.do_incr("test", vec![format!("{}", i)]);
+    }
+}
+
+#[test]
+pub fn event_action_tracker_emits_event() {
+    let threshold = 5;
+    let title = "test";
+    let message = "Exceeded tag cardinality limit";
+    let mut mock = MockClient::new();
+    let tracker_config = TrackerConfiguration::new()
+        .with_threshold(threshold)
+        .with_event(String::from(title), String::from(message));
+    let tags = vec![format!("test:{}", threshold)];
+    // Expect 100 increment calls
+    for i in 0..100 {
+        if i == threshold {
+            mock = expect_event(mock, title, &message, tags.clone());
+        }
```

Comment:
From this test it seems like it only sends the even when we reach the threshold, but I expected to be sent from that point onwards... I'm confused

Reply:
I think we only want to be sending an event or taking some action once, otherwise we end up spamming events... It seems to me like you would want to emit a warning when then threshold is crossed, and no more after that.

There are also some safety considerations - right now, the tracker drops it's state object once it's passed the threshold, which is nice from a memory and performance perspective - it ensures the tracker will never store more than `threshold` sets of tags (`Strings`), which is nice from a knowable memory impact perspective. The tag counting necessarily involves iterating across a `Vec<Set<String>>`, and it's nice to know that `Vec` has a defined upper limit on size, and as such that iteration is constrained.

Reply:
Ok, this makes sense! I don't know why I expected it to behave differently :laughing: